formats/json/jvmMain/src/kotlinx/serialization/json/internal/CharsetReader.kt

```diff
@@ -0,0 +1,124 @@
+        if (length == 0) return 0
+        require(offset in 0 until array.size && length >= 0 && offset + length <= array.size) {
+            "Unexpected arguments: $offset, $length, ${array.size}"
+        }
+
+        var offset = offset
+        var length = length
+        var bytesRead = 0
+        if (hasLeftoverPotentiallySurrogateChar) {
+            array[offset] = leftoverChar
+            offset++
+            length--
+            hasLeftoverPotentiallySurrogateChar = false
+            bytesRead = 1
+            if (length == 0) return bytesRead
+        }
+        if (length == 1) {
+            // Treat single-character array reads just like read()
+            val c = oneShotReadSlowPath()
+            if (c == -1) return if (bytesRead == 0) -1 else bytesRead
+            array[offset] = c.toChar()
+            return bytesRead + 1
+        }
+        return doRead(array, offset, length) + bytesRead
+    }
+
+    private fun doRead(array: CharArray, offset: Int, length: Int): Int {
+        var charBuffer = CharBuffer.wrap(array, offset, length)
+        if (charBuffer.position() != 0) {
+            charBuffer = charBuffer.slice()
+        }
+        var isEof = false
+        while (true) {
+            val cr = decoder.decode(byteBuffer, charBuffer, isEof)
+            if (cr.isUnderflow) {
+                if (isEof) break
+                if (!charBuffer.hasRemaining()) break
+                val n = fillByteBuffer()
+                if (n < 0) {
+                    isEof = true
+                    if (charBuffer.position() == 0 && !byteBuffer.hasRemaining()) break
+                    decoder.reset()
+                }
+                continue
+            }
+            if (cr.isOverflow) {
+                assert(charBuffer.position() > 0)
+                break
+            }
+            cr.throwException()
+        }
+        if (isEof) decoder.reset()
+        return if (charBuffer.position() == 0) -1
+        else charBuffer.position()
+    }
+
+    private fun fillByteBuffer(): Int {
+        byteBuffer.compact()
+        try {
+            // Read from the input stream, and then update the buffer
+            val limit = byteBuffer.limit()
+            val position = byteBuffer.position()
+            val remaining = if (position <= limit) limit - position else 0
+            val bytesRead = inputStream.read(byteBuffer.array(), byteBuffer.arrayOffset() + position, remaining)
+            if (bytesRead < 0) return bytesRead
+            byteBuffer.position(position + bytesRead)
+        } finally {
+            byteBuffer.flip()
+        }
+        return byteBuffer.remaining()
+    }
+
+    private fun oneShotReadSlowPath(): Int {
+        // Return the leftover char, if there is one
+        if (hasLeftoverPotentiallySurrogateChar) {
+            hasLeftoverPotentiallySurrogateChar = false
+            return leftoverChar.code
+        }
+
+        val array = CharArray(2)
```

Comment:
Probably this array can be reused?

Reply:
It will create more overhead than it reduces -- this code path is invoked very rarely